File: src/main/java/org/retrolauncher/backend/app/games/domain/entities/Game.java
package org.retrolauncher.backend.app.games.domain.entities;

import lombok.Getter;
import lombok.experimental.Accessors;
import org.retrolauncher.backend.app._shared.domain.entities.Entity;
import org.retrolauncher.backend.app.games.application.exceptions.GameValidationException;
import org.retrolauncher.backend.app.games.domain.validators.GameValidator;
import org.retrolauncher.backend.app.platforms.domain.entities.Platform;

import java.nio.file.Path;
import java.util.Optional;
import java.util.UUID;

@Getter
@Accessors(chain = true)
public class Game extends Entity {
    private String name;
    private Path path;
    private Path iconPath;
    private final UUID platformId;

    public Game(String name, Path path, UUID platformId) {
        super(UUID.randomUUID());
        this.name = name;
        this.path = path;
        this.platformId = platformId;
        validate();
    }

    public Game(UUID id, String name, Path path, Path iconPath, UUID platformId) {
        super(id);
        this.name = name;
        this.path = path;
        this.iconPath = iconPath;
        this.platformId = platformId;
        validate();
    }

    public void uploadIcon(Path path) {
        this.iconPath = path.toAbsolutePath();
    }

    public void updatePath(Path path) {
        this.path = path.toAbsolutePath();
    }

    public Optional<Path> getIconPath() {
        return Optional.ofNullable(this.iconPath);
    }

    public Game setName(String value) {
        this.name = value;
        validate();
        return this;
    }

    private void validate() {
        final GameValidator validator = new GameValidator(this);

        if (validator.hasErrors())
            throw new GameValidationException(validator.getErrors());
    }
}

File: backend/src/app/games/domain/game/game.go
package game

import (
	"retrolauncher/backend/src/app/games/domain/platform_type"

	"github.com/google/uuid"
)

type Game struct {
	id           uuid.UUID
	name         string
	platformType *platform_type.PlatformType
	cover        string
	path         string
}

func New(Name string, PlatformType *platform_type.PlatformType, Path, Cover string) *Game {
	return &Game{
		id:           uuid.New(),
		name:         Name,
		platformType: PlatformType,
		cover:        Cover,
		path:         Path,
	}
}

func Hydrate(Id uuid.UUID, Name string, PlatformType *platform_type.PlatformType, Path, Cover string) *Game {
	return &Game{
		id:           Id,
		name:         Name,
		platformType: PlatformType,
		cover:        Cover,
		path:         Path,
	}
}

func (g *Game) Update(Name string, PlatformType *platform_type.PlatformType, Path, Cover string) {
	g.name = Name
	g.platformType = PlatformType
	g.cover = Cover
	g.path = Path
}

func (g *Game) GetId() uuid.UUID                             { return g.id }
func (g *Game) GetName() string                              { return g.name }
func (g *Game) GetPlatformType() *platform_type.PlatformType { return g.platformType }
func (g *Game) GetCover() string                             { return g.cover }
func (g *Game) GetPath() string                              { return g.path }
